src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/XmlKeyHelper.cs

```diff
@@ -0,0 +1,319 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers.Binary;
+using System.Collections;
+using System.Diagnostics;
+using System.Reflection;
+using System.Text;
+
+namespace System.Security.Cryptography
+{
+    internal static class XmlKeyHelper
+    {
+        internal static ParseState ParseDocument(string xmlString)
+        {
+            if (xmlString == null)
+            {
+                throw new ArgumentNullException(nameof(xmlString));
+            }
+
+            try
+            {
+                return ParseState.ParseDocument(xmlString);
+            }
+            catch (Exception e)
+            {
+                throw new CryptographicException(SR.Cryptography_FromXmlParseError, e);
+            }
+        }
+
+        internal static bool HasElement(ref ParseState state, string name)
+        {
+            return state.HasElement(name);
+        }
+
+        internal static byte[] ReadCryptoBinary(ref ParseState state, string name, int sizeHint = -1)
+        {
+            string value = state.GetValue(name);
+
+            if (value == null)
+            {
+                return null;
+            }
+
+            if (value.Length == 0)
+            {
+                return Array.Empty<byte>();
+            }
+
+            if (sizeHint < 0)
+            {
+                return Convert.FromBase64String(value);
+            }
+
+            byte[] ret = new byte[sizeHint];
+
+            if (Convert.TryFromBase64Chars(value.AsSpan(), ret, out int written))
+            {
+                if (written == sizeHint)
+                {
+                    return ret;
+                }
+
+                int shift = sizeHint - written;
+                Buffer.BlockCopy(ret, 0, ret, shift, written);
+                ret.AsSpan(0, shift).Clear();
+                return ret;
+            }
```

Comment:
I'm not fully understanding the logic here.  What's the purpose of sizeHint?

Reply:
RSA.ImportParameters, and DSA.ImportParameters, have explicit requirements on the relationships of the sizes of the parameters.

For RSA: P, Q, DP, DQ, InverseQ must have lengths equal to half (round up) of Modulus and D must have the same length as Modulus.
For DSA: G and Y have to match P, X has to match Q.